src/v/datalake/translation/scheduling.h

```diff
@@ -0,0 +1,333 @@
+     * data.
+     */
+    virtual void notify_ready(const translator_id&) = 0;
+
+    /**
+     * Notifies that the translator with the given id has finished the currently
+     * in progress translation and has released all the resources.
+     */
+    virtual void notify_done(const translator_id&) = 0;
+
+    /**
+     * Notifies that the translators cannot make progress due to memory
+     * exhaustion.
+     */
+    virtual void notify_memory_exhausted() = 0;
+};
+
+using reservation = ssx::semaphore_units;
+
+/**
+ * Tracks the resources needed by translators. Translators work with a
+ * reservation tracker to reserve resources needed for translation. Currently
+ * tracks memory but can be extended to other resources like disk usage.
+ *
+ * note: All reservations must be destroyed before destroying the tracker.
+ */
+class reservations_tracker {
+public:
+    reservations_tracker() = default;
+    reservations_tracker(const reservations_tracker&) = delete;
+    reservations_tracker& operator=(const reservations_tracker&) = delete;
+    reservations_tracker(reservations_tracker&&) = delete;
+    reservations_tracker& operator=(reservations_tracker&&) = delete;
+    virtual ~reservations_tracker() = default;
+    /**
+     * Returns a block of memory as decided by the tracker. Can be called
+     * multiple times until the requirement is satisfied.
+     */
+    virtual ss::future<reservation> reserve_memory(ss::abort_source&) = 0;
+    /* returns true if all the memory is exhausted */
+    virtual bool memory_exhausted() const = 0;
+
+    virtual size_t allocated_memory() const = 0;
+
+    static std::unique_ptr<reservations_tracker> make_default(
+      size_t total_memory, size_t memory_block_size, scheduling_notifications&);
+};
+
+/**
+ * Current status of the translator, used as input to the scheduler.
+ */
+struct translation_status {
+    clock::duration max_target_lag;
+    size_t current_offset_lag{0};
+    // Only set if a translation is currently in progress.
+    std::optional<size_t> inflight_translated_bytes;
+    std::optional<size_t> inflight_memory_bytes_reserved;
+};
+
+std::ostream& operator<<(std::ostream&, const translation_status&);
+
+/**
+ * A single schedulable translator instance. Translation interacts with
+ * two main components
+ * - a reservation tracker - to reserve resources needed for translation
+ * - a scheduler (via scheduling_notifications) - to let the scheduler know
+ *   about the translator status.
+ *
+ * The scheduler then notifies the translator when it can begin translation
+ * via \ref translator::start_translation, along with a time slice. This allows
+ * the translator to translate for the duration of the time slice.
+ *
+ * The scheduler may also preempt the inflight translation if it chooses to
+ * (via stop_translation) and the translator should stop the inflight
+ * translation and release the resources.
+ */
+class translator {
```

Comment:
Just to clarify, the translator is an abstraction representing a long living executor of the translation that may be asked to execute a piece of work via `start_translation` ?